Feat/#55 user domain api 구현

DuckOriDuck opened the pull request:
## 🚀 개발 사항
- 회원가입, 탈퇴, 마이페이지 조회 API 구현.
- 탈퇴한 유저 조회, 존재하지 않는 유저 조회, 이미 존재하는 계정으로 가입 등에 대한 Exception 작성.
```java
USER_ACCOUNT_ALREADY_EXISTS("E01-USER-001", "이미 존재하는 계정입니다."),
USER_ACCOUNT_WITHDRAWN_ERROR("E01-USER-002", "탈퇴한 유저입니다"),

USER_ID_NOT_FOUND_ERROR("E03-USER-001", "요청된 ID를 가진 유저가 존재하지 않습니다"),
```
- Exception 핸들링 구현
### 이슈 번호
- close 55

## 특이 사항 🫶 
현재 인증 로직이 개발되지 않은 관계로 유저 정보와 인증이 필요한 API들은 우선 request parameter로 유저의 ID를 받아 작업을 수행하는 방식으로 구현해놨습니다.
```java
public ResponseEntity<MyPageAPIResponse> getMyPage(@RequestParam(required = false) Long userId/*,
                                                    @AuthenticationPrincipal UserDetails userDetails*/)
```
- Swagger는 나중에 한번에 구현해볼까 하고 생각하고 있습니다.

DuckOriDuck committed:
[FEAT] User 도메인 엔티티, 리포지토리, enum 객체 추가

DuckOriDuck committed:
[FEAT] user 정보 가져오는 메소드 추가

DuckOriDuck committed:
[FEAT] controller, service용 리스폰스 엔티티 구현

DuckOriDuck committed:
[FEAT] service id기반으로 유저의 정보를 찾아 mypage용 리스폰스 엔티티 리턴해주는 메소드 생성

DuckOriDuck committed:
[FEAT]Usercontroller 생성, mypage 신호 요청 API 구현

DuckOriDuck committed:
FIX: MyPageAPIResponse 클래스가 직렬화 가능하도록 수정

DuckOriDuck committed:
FIX: 유저 인증정보 관련 임시 주석 처리

DuckOriDuck committed:
FIX: 빌더 어노테이션 위치 수정

DuckOriDuck committed:
[FIX] user 생성자 수정, delete 날 설정하는 메소드 구현

DuckOriDuck committed:
[FEAT] user delete 기능을 위한 response DTO 생성

DuckOriDuck committed:
[FEAT] 회원탈퇴 로직 메소드 deleteUserAcount 구현

DuckOriDuck committed:
[FEAT] 회원탈퇴 API 구현

DuckOriDuck committed:
[MIGRATION] Users테이블의 pk id에 auto increment 설정

DuckOriDuck committed:
[FEAT] service의 signup을 위한 request, response dto 생성

DuckOriDuck committed:
[FEAT] 회원가입 서비스로직 구현.
에러처리 로직 필요

DuckOriDuck committed:
[FEAT] 이메일이 존재하는지 확인하는 메소드 추가

DuckOriDuck committed:
[FEAT] repository에서 save() 메소드 쓸 시 created_at 설정하는 함수 설정

DuckOriDuck committed:
[FEAT] controller API signup을 위한 request, response dto 생성

DuckOriDuck committed:
[FEAT] 회원가입 기능 /signup path 추가

DuckOriDuck committed:
[FEAT] user관련 에러 추가
"E03-USER-001", "요청된 ID를 가진 유저가 존재하지 않습니다"
"E01-USER-001", "이미 존재하는 계정입니다."
"E01-USER-002", "탈퇴한 유저입니다"

DuckOriDuck committed:
[FEAT] 탈퇴한 유저인지 체크하기 위한 utility 클래스 UserValidator 추가

DuckOriDuck committed:
[FEAT] User 엔티티에 탈퇴 취소 로직 추가

DuckOriDuck committed:
[FEAT] User Exception 처리용 클래스 추가

DuckOriDuck committed:
[FEAT] UUsreException 오타 수정

DuckOriDuck committed:
[FEAT] Email로 유저 찾도록 함수 추가

DuckOriDuck committed:
[FEAT] 이미 존재하는 회원, 존재하지 않는 회원, 탈퇴한 회원에 대한 Exception들 추가

JIN-076 commented:
오늘 중으로 `study` 도메인 api는 swagger까지 모두 작성해서 올릴 예정입니다! 제가 작성해놓은 템플릿 맞춰서 swagger 작성해주세요!👍🏻

JIN-076 commented on src/main/java/io/dev/jobprep/exception/code/ErrorCode400.java:
```diff
@@ -9,6 +9,11 @@ public enum ErrorCode400 implements ErrorCode {
     INVALID_STUDY_STATUS_TO_RECRUIT("E01-STUDY-001", "스터디가 현재 모집중이 아닙니다."),
+
+    USER_ACCOUNT_ALREADY_EXISTS("E01-USER-001", "이미 존재하는 계정입니다."),
+    USER_ACCOUNT_WITHDRAWN_ERROR("E01-USER-002", "탈퇴한 유저입니다")
```
네이밍 적을 때에 굳이 `ERROR` 키워드를 넣을 필요는 없을 것 같습니다! 

`USER_ACCOUNT_WITHDRAWN_ERROR` 보다는 `USER_ACCOUNT_DISABLED` 정도가 적당할 것 같아요! (추천)
그리고 사용자가 탈퇴를 한 경우에는 인증, 인가 권한이 없는 경우에 해당하기에 `401` 이 더 적합할 것 같습니다!

JIN-076 commented on src/main/java/io/dev/jobprep/exception/code/ErrorCode404.java:
```diff
@@ -5,6 +5,7 @@
 
 @Getter
 public enum ErrorCode404 implements ErrorCode {
+    USER_ID_NOT_FOUND_ERROR("E03-USER-001", "요청된 ID를 가진 유저가 존재하지 않습니다"),
```
이 친구도 `USER_NOT_FOUND` 가 더 깔끔할 것 같아요~

DuckOriDuck replied:
리뷰 감사드립니다. 내일 퇴근 후 전부 적용해놓도록 하겠습니다

JIN-076 commented on src/main/resources/db/migration/V3__ADD_AUTO_INCREMENT_USER_PK.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE users MODIFY id BIGINT AUTO_INCREMENT;
```
제가 지금까지 개발하면서 flyway로 version 6까지 적용해놓았습니다! 해당 sql은 V7로 변경되어야 할 것 같습니다!

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/dto/req/SignUpRequest.java:
```diff
@@ -0,0 +1,16 @@
+@AllArgsConstructor
+public class SignUpRequest {
+    private String email;
+    private String username;
```
Dto 클래스는 불변해야 하는 클래스이니 final 키워드를 사용하면 좋을 것 같습니다!

JIN-076 commented on src/main/java/io/dev/jobprep/users/domain/User.java:
```diff
@@ -0,0 +1,76 @@
+import java.time.LocalDateTime;
+
+@Getter
+@NoArgsConstructor
```
@NoArgsConstructor를 사용할 때는 `access = AccessLevel.PROTECTED` 로 접근 제한을 해서 기본 생성자의 무분별한 생성을 방지하는 것이 좋을 것 같습니다!

JIN-076 commented on src/main/java/io/dev/jobprep/users/domain/User.java:
```diff
@@ -0,0 +1,76 @@
+    @Column(name="penalty_updated_at")
+    private LocalDateTime penaltyUpdatedAt;
+
+    @CreationTimestamp
```
제가 common/base 패키지에 만들어놓은 `BaseTimeEntity` 클래스를 상속하면 `createdAt` `deletedAt` 클래스를 명시해주지 않아도 됩니다~!

JIN-076 commented on src/main/java/io/dev/jobprep/users/domain/User.java:
```diff
@@ -0,0 +1,76 @@
+        this.email = email;
+        this.userRole = userRole;
+        this.penaltyUpdatedAt = penaltyUpdatedAt;
+        this.createdAt = createdAt;
```
`BaseTimeEntity` 클래스를 상속하면 @CreatedDate가 붙어있어서 해당 객체가 데이터베이스에 저장되는 과정에서 자동으로 현재 시간을 매핑해줍니다! 그래서 생성자에서 `createdAt` `deletedAt` 을 인자로 받을 필요가 없습니다!

그러면 onCreate(), setDeletedAt() 모두 지우셔도 돼요!

JIN-076 commented on src/main/java/io/dev/jobprep/users/domain/User.java:
```diff
@@ -0,0 +1,76 @@
+        return now;
+    }
+
+    public void unSetDeletedAt() {
```
메서드명이 직관적이면 좋을 것 같아요! '`deletedAt` 을 undo할거야' 보다는 '재가입할거야' 라는 의미의 `rejoin` 혹은 '복구' 라는 의미의 `restore` 가 직관적이어서 코드 가독성에 좋을 것 같아요!

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/UserService.java:
```diff
@@ -0,0 +1,84 @@
+                .build();
+
+        //TODO error exception 처리
+        return SignUpResponse.from(userRepository.save(newUser));
```
해당 Dto에 대해서는 정적 팩토리 메서드를 사용하셨는데 `MyPageResponse` `DeleteUserAccountResponse` 는 빌더 패턴을 사용하신 이유가 있을까요? 다 같이 정적 팩토리 메서드 네이밍을 사용하면  통일성이 있어보일 것 같습니다!

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/UserService.java:
```diff
@@ -0,0 +1,84 @@
+            User existingUser = userData.get();
+
+            try {//탈퇴했던 회원인지 체크
+                UserValidator.validateUserActive(existingUser);
```
Validator 클래스를 별도로 두셨군요! 이것도 좋은 방법입니다!

저희가 기본 컨셉으로 가지고 가기로 했던 도메인 주도 설계의 관점으로 본다면, 해당 로직을 통해 검증하고 싶은 내용은 '유저가 탈퇴를 했는가'이기 때문에 user 도메인에 있는 `deletedAt` 필드와 밀접하게 관련이 있습니다. 그렇기 때문에 해당 유저가 현재 탈퇴한 유저인지를 검증하는 코드는 User 클래스에 다음과 같이 들어가고 사용되는 것이 좋을 것 같습니다.

```java
public class User {

    ...
    private LocalDateTime deletedAt;

    public void validateUserActive() {
       if (this.deletedAt != null) {
          throw new UserException("~~");
       }
     }

}

public class UserService {

   @Transactional
   public SignUpResponse signUp(SignUpRequest signUpRequest) {

      Optional<User> user = userRepository.findUserByEmail(signUpRequest.getEmail());
      ...

      try {
         user.validateUserActive();
      } catch(...) {
         ...
      }
      
       ...
   }
}
```

다음 링크 참고하시면 어떤 느낌인지 오실 것 같아요.
[아키텍처 개요](https://velog.io/@andy230/%EC%95%84%ED%82%A4%ED%85%8D%EC%B2%98-%EA%B0%9C%EC%9A%94)
[애그리거트](https://velog.io/@andy230/Aggregate-%EC%95%A0%EA%B7%B8%EB%A6%AC%EA%B1%B0%ED%8A%B8)

DuckOriDuck replied:
음~ 에러 처리 로직은 서비스에만 몰아가야 한다고 생각해서 따로 validator 클래스를 둔 것인데 그냥 엔티티단에서 throw하고 service단에서 받는 방식으로 처리해도 상관 없군요. 많이 고민했던 부분인데 상세하게 적어주셔서 감사합니다

JIN-076 commented on src/main/java/io/dev/jobprep/users/presentation/UserController.java:
```diff
@@ -0,0 +1,47 @@
+        return ResponseEntity.ok(MyPageAPIResponse.from(myPageResponse));
+    }
+
+    @DeleteMapping("/delete-account")
```
HTTP 메서드인 `DeleteMapping` 을 사용함으로써 해당 api 엔드포인트가 어떤 작업을 하려고 하는지 의미한다고 생각합니다. 좀 더 RESTful하게 작성한다면 `DELETE /api/v1/users` 로 충분할 것 같습니다.

JIN-076 commented on src/main/java/io/dev/jobprep/users/infrastructure/UserRepository.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.Optional;
+
```
여기에 @Repository 명시해주세요~

JIN-076 commented on src/main/java/io/dev/jobprep/users/domain/User.java:
```diff
@@ -0,0 +1,76 @@
+    protected void onCreate() {
+        this.createdAt = LocalDateTime.now();
+    }
+    public User(Long id,
```
Builder를 사용해서 `User` 객체를 생성하고 있으니, 요 생성자는 private으로 무분별한 생성을 막으면 좋을 듯 합니다~

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/UserService.java:
```diff
@@ -0,0 +1,84 @@
+        User userData = userRepository.findUserById(userId)
+                .orElseThrow(() -> new UserException(USER_ID_NOT_FOUND_ERROR));
+        //이미 탈퇴처리 된 유저 처리
+        UserValidator.validateUserActive(userData);
```
이렇게 작성하면 훨씬 직관적으로 도메인 모델을 작성할 수 있습니다! 가독성이 확실히 증가하게 되더라구요👍🏻

```java
public DeleteUserAccountResponse delete(Long userId) {

   User user = userRepository.findUserById(userId).orElseThrow(() -> new UserException("~~"));
   user.delete();
}

public class User extends BaseTimeEntity {

   public void delete() {
      validateAvailableDelete();
      delete(); // for BaseTimeEntity
   }

   public void validateAvailableDelete() {
     if (getDeletedAt() != null) {
        throw new UserException("~~");
     }
   }
}

@Getter
public class BaseTimeEntity {

   protected void delete() {
      this.deletedAt = LocalDateTime.now();
   }
}
```

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/UserService.java:
```diff
@@ -0,0 +1,84 @@
+
+        LocalDateTime timeNow = userData.setDeletedAt();
+
+        return DeleteUserAccountResponse.builder()
```
사용자가 '탈퇴하시겠습니까?' 질문에 '탈퇴하기' 버튼을 누르고 나면 처음 화면으로 돌아가기 때문에 굳이 탈퇴한 시간과 메시지를 전달할 필요가 있을까요? 저희는 클라이언트 측에 탈퇴가 성공했음을 알리는 200 상태코드만 알리면 클라이언트 측에서 alert로 메시지를 띄워줄 것 같습니다!

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/UserService.java:
```diff
@@ -0,0 +1,84 @@
+            try {//탈퇴했던 회원인지 체크
+                UserValidator.validateUserActive(existingUser);
+            }catch(UserException e){//탈퇴취소 로직
+                existingUser.unSetDeletedAt();
```
저희 피그마에서 회원 탈퇴 페이지를 보면 '계정을 삭제하시면 모든 정보가 삭제되며, 복구할 수 없습니다' 라고 명시가 되어있는데 재가입은 가능한걸까요? 복원은 안되지만 재가입은 된다,,? 지금 로직대로면 재가입이 곧 복원같은데 이건 기획팀이랑 얘기 맞춰봐야 할 것 같아요~

DuckOriDuck replied:
일단 냅두겠습니다

JIN-076 commented:
저는 개발하면서 코드를 가독성, 유연성 있게 작성하는 데에 주의를 많이 들이는 편입니다! 그러다 보니 피드백이 너무 많아졌는데 꼭 필수적으로 수정해야 하는 것은 아닙니다! 알아두시면 나중에 리팩토링하거나 클린 코드 작성하실 때 좋으실 것 같습니다!

DuckOriDuck committed:
[FIX] 에러코드 이름 변경, USER_ACCOUNT_DISABLED 에러 401로 이동

DuckOriDuck committed:
[RENAME] 마이그레이션 파일 버젼 변경 V3->V7

DuckOriDuck committed:
[REFACTOR] DTO 클래스들에 final 추가

DuckOriDuck committed:
[FIX] 전부 다 이미 존재하는 계정으로 감지하는 에러 수정

DuckOriDuck committed:
[FEAT] 탈퇴 취소 로직 추가

DuckOriDuck committed:
[FIX] 유저 코드 수정
- 생성자 수정
- 상속 적용되도록 변수 수정
- 탈퇴/탈퇴취소 로직 수정

DuckOriDuck committed:
[REFACTOR] unSetDelete->restore, resotoreAccount()로 변경

DuckOriDuck committed:
[REFACTOR] 정적 팩토리 메소드로 변경

DuckOriDuck committed:
[REFACTOR] Validator 클래스 삭제, error throw는 유저단에서 하도록 설정

DuckOriDuck committed:
[FIX] value change를 Super class에서 해버리면 JPA가 값의 변경을 인지하지 못해서 DB에 write가 되지 않음.
write가 되게 하려면 repository를 사용해서 save하거나 entity 단에서 바꾸던가 해야하는데 entity에서 repository.save() 하는 것은 좋지 않은 방법 같아 BaseTimeEntity의 필드 변수들의 접근 제한을 protected로 설정.
- delete()는 혹시 모르니 놔두는 용

DuckOriDuck committed:
[REFACTOR] DeleteMapping path 수정

DuckOriDuck committed:
[FIX] repository 명시

DuckOriDuck committed:
[FIX] @Transactional 붙이니 해결됨.
setDelete LocadateTime 리턴하도록 설정

DuckOriDuck committed:
[FIX] 탈퇴시간 리턴 안하도록 수정
